functions/index.js

```diff
@@ -95,8 +96,112 @@ exports.TrackUsers = functions.https.onRequest(async (req, res) => {
     });
 });
 
-function logUnfetchedCourse(course, roster) {
-  console.log(`Unable to fetch course data for course: ${course} in roster: ${roster}`);
+function typeToOrderedNumber(type) {
+  switch (type) {
+    case 'WI':
+      return 0;
+    case 'SP':
+      return 1;
+    case 'SU':
+      return 2;
+    case 'FA':
+      return 3;
+    default:
+  }
+}
+
+function compareDataObjectsByRosters(dataObject1, dataObject2) {
+  const roster1 = dataObject1.roster;
+  const roster2 = dataObject2.roster;
+  const type1 = roster1.slice(0, 2);
+  const year1 = roster1.slice(2);
+  const type2 = roster2.slice(0, 2);
+  const year2 = roster2.slice(2);
+
+  if (year2 > year1) {
+    // roster2 has more recent year than roster1
+    return 1;
+  } else if (year1 > year2) {
+    // roster1 has more recent year than roster2
+    return -1;
+  } else if (typeToOrderedNumber(type2) > typeToOrderedNumber(type1)) {
+    // roster2 has more recent semester type than roster1
+    return 1;
+  } else if (typeToOrderedNumber(type1) > typeToOrderedNumber(type2)) {
+    // roster1 has more recent semester type than roster2
+    return -1;
+  } else {
+    return 0;
+  }
+}
+
+function sortDataCrseInfoByMostRecentRosters(dataCrseInfo) {
+  // Sorts from most recent roster
+  return dataCrseInfo.sort(compareDataObjectsByRosters);
+}
+
+function logUnfetchedCrseCode(crseCode, roster) {
+  console.log(`Unable to fetch course data for crseCode: ${crseCode} in roster: ${roster}`);
+}
+
+function logUnfetchedCrseId(crseId, roster) {
+  console.log(`Unable to fetch course data for crseId: ${crseId} in roster: ${roster}`);
+}
+
+function logInvalidRoster(roster) {
+  console.log(`Inputted roster: ${roster} is invalid`);
+}
+
+function cleanDataCrseInfoByRoster(dataCrseInfo) {
+  let cleanedDataCrseInfo = [];
+  dataCrseInfo.forEach(dataObject => {
+    let roster = dataObject.roster;
+    if (typeof roster === 'string' && roster.length === 4) {
+      roster = roster.toUpperCase();
+
+      let type = roster.slice(0, 2);
+      const validTypes = ['WI', 'SP', 'SU', 'FA'];
+
+      let year = roster.slice(2);
+      if (!(validTypes.includes(type) && parseInt(year, 10))) {
```

Comment:
I see that this checks if it is a valid season/year, but should we check if we actually have data for that season/year? Might be unnecessary idk

Reply:
In the case where the roster is well formed but we might not have the roster info ("SP 30"), we just wouldn't fetch any courses for that roster. I think it's ok to leave as is for now because I think the main problem was how to handle malformed rosters (i.e. 1232, "hihoih") which would break how we compare rosters and use `toUpperCase()`.